### DISCUSSION TITLE
I-Want-To-Contribute Thread 🚀

### DISCUSSION THREAD

**Discussion by tomhea** [tomhea/flip-jump#148] [Ideas]
Hello all!

If you want to contribute or have any ideas, feel free to share them here (or in any other conversation) 😄 

Tag me freely,
Tom

**Comment by tomhea** [comment]
**Contribution Ideas**:

- Writing a cool FlipJump program (preferably with [test support](https://github.com/tomhea/flip-jump/blob/main/tests/README.md))
- Adding a feature to the [standard library](https://github.com/tomhea/flip-jump/tree/main/stl)
- Writing a new standard library module
- Optimize an stl macro
- Testing a feature
- Suggesting a cool new addition to the FlipJump assembly 
- Suggesting a better way of preserving information (documentation, etc.)
- Optimize the interpreter
- Optimize the assembler
- Writing an interpreter in a different language (kinda easy, you only need to know the [.fjm format](https://github.com/tomhea/flip-jump/blob/main/src/fjm.py))
- Writing an assembler in a different language (more of a challenge)
- Writing (/refactoring to) a cleaner and better code

Discuss here before you start to hear some insight and tips, and to make sure no one else is already working on it 😉

**Reply by tomhea** [reply]
@lestrozi you are most welcome here

**Reply by lestrozi** [reply]
Hey, "Optimize a" is missing the end :)

I'm not very familiar with the stdlib, but if you think it's lacking something interesting, I could try to contribute

Honestly, I wouldn't worry too much about refactoring/improving/rewriting in another language the interpreter/assembler (unless someone thinks this is fun.. it could even be a CS exercise on compilers/lexers).

As for documentation, one thing that took me a while to figure out was how to make a counter (one that actually increments a variable, not one that repeat the code). I could write a doc on that (unless there's something already that I missed?).

Another cool project could be to add a "graphical interface" to it (perhaps a 320x240 memory address that gets converted to a monochromatic image). This could be used to support coding a Pong - and thinking long-term, maybe a real hardware (FPGA) that implements a flip-jump processor and can execute Pong on a TV :P - but that's.. a long project

**Reply by tomhea** [reply]
Not sure how much you read, but a lot of the stdlib information is on the esolang page.
For example - this is the [list](https://esolangs.org/wiki/FlipJump#The_Standard_Library) of the useful macros in the stdlib. 
most macros work on [bits](https://github.com/tomhea/flip-jump/blob/main/stl/bitlib.fj) (are you familiar with [the concept](https://esolangs.org/wiki/FlipJump#Memory_-_how_can_we_implement_variables.3F)?)

So for the counter, you can count do:
```
hex.inc 4, counter  // counter = 1
hex.inc 4, counter  // counter = 2
hex.print 4, counter  // prints "0002"
...
end: ;end
counter: hex.vec 4, 0
```

But I feel you are talking about a different type of counter. If you are talking about a way to increment the *flip** / **jump** part of an opcode by 1 - I don't have documentation of it. I kinda did it [here](https://github.com/tomhea/flip-jump/blob/87332206a3d807854fb9a4d0abd8fd232cbeb99e/stl/ptrlib.fj#L101), by advancing the .to_flip flip-part.
So documentation of what you did can be very useful.
I'd say a good place to put it is in docs/YourDocName.md.
If you want to pull request, ask to merge your branch to dev, and from there I'll merge it myself in the next version of main.

The standard library isn't documented enough (or at all in the code itself) in my opinion. I'd like you to take a look and try to tell me where can I improve so that it will be easy for new programmers to enter and start using the standard library. Anything you think that could help that goal, I'm open :) I also want you and everyone to not miss the macros I already programmed that are in the stdlib.

As for the graphical interface, you got me very interested. And I have some ideas! The pong sounds awesome.
Let's talk about it over [here](https://github.com/tomhea/flip-jump/discussions/152) (a new discussion).

**Reply by lestrozi** [reply]
Yeah, I eventually got some familiarity with the stdlib, but initially I read only the main doc and it was hard to me to figure out how to increment a counter. With the stdlib indeed it's easy, but.. what got me interesting in FJ is how simple it is and how it can make complex things, so just using stdlib was no fun as I'd basically be programming in a "normal" language.

I kinda explain on quine16.fj how I did the counter, so TBH I'm also more interested on the GUI :D
I'll have a look at the other discussion.

**Reply by tomhea** [reply]
I can really relate to your low-level interest, flipping the bits myself and see what real difference it makes is much more fun. 😀 

I indeed have one low-level mission I didn't finish and it might just be for you.
The mission is Implementing integer division for `hex` variables (which are intended to be faster than the old `bit` variables). Basically, writing the fast division every fj program should use. If you are interested in making a change and influence the stdlib itself - and you are math-oriented - that can be a great mission for you. 
If you are interested let me know!

I'm going to work in the next days on documenting and making the stl more accessible, so I can adjust myself and start with might be relevant for this project.

**Reply by lestrozi** [reply]
Am I right that there already is a hex.div (I remember seeing some test)? Is the idea you're suggesting to make it more efficient?

**Reply by tomhea** [reply]
I did write a `hex.div`, but it currently don't pass any of my tests ([programs/hexlib_tests/div/](https://github.com/tomhea/flip-jump/tree/main/programs/hexlib_tests/div)). 
This macro performs unsigned integer division. I chose to implement it by repeated subtraction.
The arguments are: 
```
   n - length of _a
  nb - length of _b
   q - the _a/_b (division) result will be written here
  _r - the _a%_b (reminder) result will be written here
  _a - the unsigned number to divide
  _b - the unsigned number to be divided by
div0 - address to jump if _b is 0.
```
You can run the div tests by yourself, with `pytest --hexlib --contains div -svv` (main branch, from the project's root directory).

You can see the tests arguments in [compile_csv](https://github.com/tomhea/flip-jump/blob/main/tests/test_compile_hexlib.csv#L10), and [run_csv](https://github.com/tomhea/flip-jump/blob/main/tests/test_run_hexlib.csv#L10).
If you are not familiar, the test-csv format can be found [here](https://github.com/tomhea/flip-jump/blob/main/tests/README.md#compile-csvs-format).

By the way, hex.div suppose to be the improved version of bit.div and bit.div_loop - using the O(wn) space from div_loop (which is the smaller of the two), and being faster than bit.div (which is the faster one of the two).
The improvement is mainly using the fcall (fast-call), and the hex variable.

**Reply by tomhea** [reply]
> Am I right that there already is a hex.div (I remember seeing some test)? Is the idea you're suggesting to make it more efficient?

@lestrozi 
BTW, I know that a lot of time past since, but I run into your comment so just updating that hex.div works and it's much more efficient :)

**Reply by tomhea** [reply]
And I remember right now that we didn't do any gui at the end here

**Reply by lestrozi** [reply]
Hey there! Nice to see you're still interested in this project!

Not sure if I mentioned, but I have implemented a simple VM that can do very basic operations (including drawing): https://github.com/lestrozi/pipevm/

Shouldn't be too difficult to convert the 2 brainfck projects there (pong and shell) to fj (I suppose you have a converter, no?). It will probably be very inefficient though, perhaps unplayable :P but worth a shot!

**Reply by tomhea** [reply]
Oh nice! Looks pretty cool.
I do have a bf2fj compiler :) nice guess
https://github.com/tomhea/bf2fj

What are you working on these days?

**Reply by lestrozi** [reply]
I've been spending most of my time working professionally with Rust+Linux on some low-level networking, but I don't have any pet projects atm. Up for suggestions :)

**Comment by lestrozi** [comment]
Btw, after my comment about interrupt/syscalls in the other discussion, another idea would be to implement some of BIOS interrupts (as you mentioned, we'd need something similar to int 16h to check/read key presses, as well as int 10h for graphics).

But I'm not very experienced in firmwares to be very useful

**Reply by tomhea** [reply]
Answered in #154
